### ISSUE TITLE
Jerome's Checklist

### ISSUE THREAD

**Issue by Ishaan-Datta** [UBCAgroBot/ROS#4]
- [ ] Converting .onnx file to .trt in colab
- [ ]  Implementing the TensorRT engine inference using NVIDIA examples and other tutorials in C++
- [ ]  Optimizing the inference with batching, CUDA contexts, multithreading/asynchronous streams
- [ ]  Include best practices: [NVIDIA TensorRT optimizations](https://docs.nvidia.com/deeplearning/tensorrt/archives/tensorrt-700/tensorrt-best-practices/index.html) (include creating interfaces or abstract classes that define the common functionality required for neural network processing)
- [ ]  Comparison w/ industry examples (Nvidia, Zed, Github examples) to adapt our case to the Jetson hardware
- [ ]  Metric logging for benchmarking using Nvidia tutorials
- [ ]  Test different batch sizes to maximize the GPU utlilization
- [ ]   Investigate any other potential areas for multiprocessing or parallel execution to speed up task processing

Links Checklist:

- [ ]  [TensorRT C++ API Repository](https://github.com/cyrusbehr/tensorrt-cpp-api?tab=readme-ov-file)
- [ ]  [Youtube Tutorial](https://www.youtube.com/watch?v=Z0n5aLmcRHQ)
- [ ]  [Nvidia Support Matrix](https://docs.nvidia.com/deeplearning/tensorrt/support-matrix/index.html#fntarg_26)
- [ ]  [Nvidia Archived Best Practices](https://docs.nvidia.com/deeplearning/tensorrt/archives/tensorrt-700/tensorrt-best-practices/index.html)
- [ ]  [Nvidia Object Detection Examples](https://docs.nvidia.com/deeplearning/tensorrt/sample-support-guide/index.html#object-detection)
- [ ]  [Nvidia Guide for Benchmarking/Profiling](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#measure-performance)
- [ ]  [Nvidia Guide for Optimization](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#optimize-performance)
- [ ]  [Nvidia Example ONNX Deployment](https://docs.nvidia.com/deeplearning/tensorrt/quick-start-guide/index.html#ex-deploy-onnx)
- [ ]  [Nvidia Basic Workflow (for making docs)](https://docs.nvidia.com/deeplearning/tensorrt/quick-start-guide/index.html#basic-workflow)

Links Checklist (some of these may overlap w/ Jerome's so maybe meet with him to discuss progress and what needs to be revisited or reviewed)

- [ ]  [YOLOV8](https://github.com/cyrusbehr/YOLOv8-TensorRT-CPP)
- [ ]  [ZedBox Custom Input](https://github.com/stereolabs/zed-sdk/tree/master/object%20detection/custom%20detector/cpp/tensorrt_yolov5-v6-v8_onnx)
- [ ]  [ONNX TensorRT](https://github.com/onnx/onnx-tensorrt/tree/main)
- [ ]  [Python Tutorial Script](https://freedium.cfd/https://alimustoofaa.medium.com/how-to-load-model-yolov8-tensorrt-6c03de059622)
- [ ]  [Nvidia Tutorial Example](https://developer.nvidia.com/blog/speeding-up-deep-learning-inference-using-tensorflow-onnx-and-tensorrt/)
- [ ]  [Updated Nvidia Tutorial](https://developer.nvidia.com/blog/speeding-up-deep-learning-inference-using-tensorrt-updated/)
- [ ]  [TRTexec Tool Script](https://github.com/triple-Mu/YOLOv8-TensorRT/blob/main/docs/Normal.md)
